src/main/java/appeng/util/item/AEItemStack.java

```diff
@@ -123,14 +133,40 @@ public static IAEItemStack fromPacket( final ByteBuf data ) throws IOException
 			return null;
 		}
 
-		final AEItemStack item = AEItemStack.fromItemStack( itemstack );
-		item.setStackSize( stackSize );
+		AEItemStack item = null;
+		if( Platform.isServer() && stackId > 0 )
+		{
+			final AESharedItemStack existing = AEItemStackRegistry.getRegisteredStack( stackId );
+			if( existing != null )
+			{
+				item = new AEItemStack( existing, stackSize );
+			}
+		}
+		if( item == null )
+		{
+			item = new AEItemStack( AEItemStackRegistry.getRegisteredStack( itemstack, stackId ), stackSize );
+		}
+
 		item.setCountRequestable( countRequestable );
 		item.setCraftable( isCraftable );
 		return item;
 	}
 
 	@Override
+	public void writeToPacket( final ByteBuf i ) throws IOException
+	{
+		final byte mask = (byte) ( this.getType( this.getSharedStack().getStackId() ) | ( this.getType( this.getStackSize() ) << 2 ) | ( this
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Prevent "int" promotion by adding "& 0xff" to this expression. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS3034)